### PR TITLE
OpenCV renderer replacing PyTorch3D as the default

### PR DESCRIPTION
**Pull request by adscib**
This PR adds a cv2 renderer and makes pytorch3d an optional dependency. That means people will be able to pip-install a fully functional version of torchdrivesim without any extra hassle. The new renderer is CPU-only, and it seems to run fast enough to render the view of one agent in real time, but it is sequential, so it will be quite slow for multiple ego agents or batches.

I extended tests to cover all the different backends, and I confirmed that the package installs correctly with pip in a fresh environment, and that the tests pass as expected (skipping those that depend on either pytorch3d or nvdiffrast). Below I show some comparisons of an example frame with different renderers:

cv2
cv2
cv2_egocentric

pytorch3d
pytorch3d
pytorch3d_egocentric

nvdiffrast
nvdiffrast
nvdiffrast_egocentric

### CODE REVIEW

**Comment by adscib** [comment]
@lioutasb do you think it would be possible to make the nvdiffrast renderer work without pytorch3d? It currently uses some functions from it, and I'm not sure how difficult it would be to reimplement them.

**Comment by adscib** [comment]
@Ruishenl what do you think we should do about the CI? It would be nice to test both that everything works without pytorch3d and nvdiffrast, and also that they work correctly when available.

**Comment by lioutasb** [comment]
> @lioutasb do you think it would be possible to make the nvdiffrast renderer work without pytorch3d? It currently uses some functions from it, and I'm not sure how difficult it would be to reimplement them.

Yeah, it should be doable. I have experience with the inner workings of the pytorch3d cameras which is really what is needed from pytorch3d (and some batched mesh manipulation).

**Comment by Ruishenl** [comment]
> @Ruishenl what do you think we should do about the CI? It would be nice to test both that everything works without pytorch3d and nvdiffrast, and also that they work correctly when available.

I can add a new test workflow in a different environment (Dockerfile) without pytorch3d and nvdiffrast.

File: torchdrivesim/rendering/base.py [hunk 1]
```diff
@@ -42,6 +40,35 @@ class DummyRendererConfig(RendererConfig):
     backend: str = 'dummy'
 
 
+class Cameras:
+    """
+    Lightweight version of pytorch3d.renderer.FoVOrthographicCameras.
+    Used to define an interface that works without pytorch3d.
+    """
+    def __init__(self, xy: Tensor, sc: Tensor, scale: float):
+        self.xy = xy
```
Comment (Ruishenl) [inline torchdrivesim/rendering/base.py:49]: Should we enforce dimension for passed tensors and also keep track of the device used for this class ?

**Comment by adscib** [comment]
> I can add a new test workflow in a different environment (Dockerfile) without pytorch3d and nvdiffrast.

Go for it.

**Comment by adscib** [comment]
> Yeah, it should be doable. I have experience with the inner workings of the pytorch3d cameras which is really what is needed from pytorch3d (and some batched mesh manipulation).

If it's not too much effort, can you include it in this PR? It's not crucial, so we could skip it for now if it would take too much work. Similarly for the offroad infraction, which currently relies on pytorch3d.

File: torchdrivesim/rendering/base.py (hunk 1 above)
Comment (adscib) [reply torchdrivesim/rendering/base.py:49]: I think I would prefer to keep it lightweight for now. I was thinking we might need to add some coherent shape checks throughout the library, and if we do, we could address this at the same time.

**Comment by lioutasb** [comment]
> > Yeah, it should be doable. I have experience with the inner workings of the pytorch3d cameras which is really what is needed from pytorch3d (and some batched mesh manipulation).
> 
> If it's not too much effort, can you include it in this PR? It's not crucial, so we could skip it for now if it would take too much work. Similarly for the offroad infraction, which currently relies on pytorch3d.

Yeah I can remove the pytorch3d dependency from nvdiffrast. I'm not that confident about the offroad infraction though as I only have a high level idea of `point_face_distance` works (we can deal with that in a different PR).

**Comment by lioutasb** [comment]
I changed `nvdiffrast` to not rely on `pytorch3d` cameras. I believe it works correctly but for some reason I can't run `pytest` locally. Someone should confirm that the renderings look correct before merging.

**Comment by lioutasb** [comment]
I'm looking at the offroad loss now. It might be possible to do the computation using pytorch ops only (but I don't know yet how much slower it will be compared to pytorch3d).

**Comment by adscib** [comment]
> I changed `nvdiffrast` to not rely on `pytorch3d` cameras. I believe it works correctly but for some reason I can't run `pytest` locally. Someone should confirm that the renderings look correct before merging.

Thank you! I have confirmed that the nvdiffrast renderer is still working correctly, and that it runs when pytorch3d is not available.
nvdiffrast
nvdiffrast_egocentric

**Comment by lioutasb** [comment]
I added a pytorch only offroad loss. The tests are passing because the code returns zero due to the empty road mesh. We should change the tests and test with a non-empty road mesh to see if it doesn't fail. My intention here is to use pytorch3d whenever it's possible and default back to the custom implementation if it's not installed. I tested the output of the two methods and it looks identical in all of my tests but you may want to do further comparisons as well.

**Comment by adscib** [comment]
Perfect! Are we expecting an exact equivalence or just approximate?

**Comment by lioutasb** [comment]
Exact equivalence (subject to small numerical precision errors) since I based it on their pytorch cuda code. I'm almost certain that it's also differentiable but I haven't tried it yet. I'll test it before we merge so we can update the docs. I believe my implementation is more memory intensive and somewhat slower since it calculates many unnecessary operations that you would be able to skip when doing things in low-level cuda.

**Comment by adscib** [comment]
We can set up some training jobs to run both methods and check for equivalence. That will probably be the most efficient way to test it. Since this change doesn't break any existing code, I think we can go ahead and merge. The PR is finished as far as I'm concerned.

**Review by lioutasb — APPROVED** [review]
